crypto/bn/bn_rand.c

```diff
@@ -36,8 +36,8 @@ static int bnrand(BNRAND_FLAG flag, BIGNUM *rnd, int bits, int top, int bottom,
     if (bits < 0 || (bits == 1 && top > 0))
         goto toosmall;
 
-    bytes = (bits + 7) / 8;
     bit = (bits - 1) % 8;
+    bytes = bits / 8 + (14 - bit) / 8;  /* Same as (bits + 7) / 8 but can't overflow */
```

Comment:
This exceeds the limit in bn_expand, so
in the end the BN_bin2bn function will fail.

Reply:
... And the limit in bn_expand_internal is even lower.  Around INT_MAX/4.
This tries out where the effective limit is:
```C
#include <openssl/bn.h>
#include <stdio.h>

int main()
{
  BIGNUM *bn;
  int res;
  bn = BN_new();
  res = BN_rand(bn, INT_MAX / 4 - BN_BITS2 + 2, 0, 0);
  printf("res=%d\n", res);
  res = BN_rand(bn, INT_MAX / 4 - BN_BITS2 + 2, BN_RAND_TOP_ANY, 0);
  printf("res=%d\n", res);
  res = BN_rand(bn, INT_MAX / 4 - BN_BITS2 + 1, 0, 0);
  printf("res=%d\n", res);
  return 0;
} 
```
the result is for me
first test: res=0 with 100%
second try: res=0 with 50% and res=1 with 50%
third test: res=1 with 100%

So I would suggest have a hard limit at INT_MAX/4 since it is
very unlikely that so BN_BITS2-1 random top bits will be 0.

Reply:
@tom-cosgrove-arm will you update this PR according to the above comment?

Reply:
ping @tom-cosgrove-arm ^^

Reply:
Yes, will do, have been on leave and now catching up

Reply:
ping @tom-cosgrove-arm